### PR TITLE
Fix #831

### LINKED ISSUE #831: Bug in reader: can't terminate string literals containing \ at end of line.
Common Lisp had docstrings before Python existed. Hy should totally support this.
```Python
>>> def foo():
	"""I'm a docstring!"""
	return 42

>>> help(foo)
Help on function foo in module __main__:

foo()
    I'm a docstring!

>>> foo.__doc__
"I'm a docstring!"
```
Let's try it in Hy.
```
=> (defn foo ()
... """I'm a docstring"""
... 42)
=> (help foo)
Help on function foo:

foo()

=> foo.__doc__
''
```
?!
Where's my docstring?
(some time later...)
Let's try hy2py.
```Python
def foo():
    ''
    "I'm a docstring!"
    ''
    return 42
```
I see what you did there, Hy. Yeah, there should probably be a warning about that somewhere. Triple  quotes don't work. Why does Python use them anyway? So we can include newlines mostly. Common Lisp strings do that already. Does Hy?
```
=> "one
... two"
'one\ntwo
```
Yup. Looks good.
Escapes also appear to work. But what if I actually wanted backslashes? Do raw strings work?
```
=> r"\foo"
'\\foo
```
Nice. Now **raw** triple quoted strings? This works in Python:
```Python
>>> r"""\
foo"""
'\\\nfoo'
```
Let's use what we've learned and try it in Hy.
```
=> r"\
... foo"
... "
... ""
... """
... what?
... \
... \
... \\
... \\\
... \\\\
... \"
... \\"
... \\\"
... \\\\"
...   C-c C-c

KeyboardInterrupt
```
Something is broken...

### PR DESCRIPTION
**Pull request by kirbyfan64**
Well, this was *super* hard. It took me a whole 60 seconds!! ;)

In reality, writing the tests took the longest because I kept messing up the slashes.

@gilch Approve?

### CODE REVIEW

**Comment by gilch** [comment]
>  tests took the longest because I kept messing up the slashes.

Python's raw strings should help with that. The test would be a lot more readable as:
```Python
tokenize(r'''
r"a\
bc"
''')
```
This way you can see the intended Hy exactly as written. Future readers of the tests won't have to be as confused. Also a comment mentioning the `DOTALL` flag in Python's `re` might better clarify the purpose of the change.

**Comment by kirbyfan64** [comment]
@gilch Done.

File: tests/lex/test_lex.py [hunk 1]
```diff
@@ -98,8 +98,14 @@ def test_lex_symbols():
 
 def test_lex_strings():
     """ Make sure that strings are valid expressions"""
-    objs = tokenize("\"foo\" ")
+    objs = tokenize('"foo"')
     assert objs == [HyString("foo")]
+    # See issue #831.
```
Comment (zackmdavis) [inline tests/lex/test_lex.py]: I think this comment could be more descriptive, like `#  can tokenize (ignored) backslash-newline`. (Source comments occupy valuable real estate; dedicated future readers who want to find an associated ticket can always look for more info in the commit message, but casual readers might be annoyed by the exhortation to look up a magic number on GitHub.)

**Comment by zackmdavis** [comment]
***+1***

File: tests/lex/test_lex.py (hunk 1 above)
Comment (kirbyfan64) [inline tests/lex/test_lex.py]: @zackmdavis Fixed.

**Comment by gilch** [comment]
:+1:

**Comment by kirbyfan64** [comment]
So, I feel *really* stupid right now, because in the original comment I wrote:

> This took me a whole 60 minutes!!

It was supposed to be a joke; I meant to write:

> This took me a whole 60 seconds!!

My programming skills aren't awful enough to make this take an hour... :(

**Comment by algernon** [comment]
:+1:

**Pull request merged**